### ISSUE TITLE
Per package/component `cabal clean`

### ISSUE THREAD

**Issue by michaelpj** [haskell/cabal#7506]
`cabal clean` is both simple and effective, by virtue of just deleting everything. However, in a multi-package project, I usually want to clean just one package or component that I'm working on. 

Of course, the current implementation works for this, but once you accrue a few local packages and a few `source-repository-package`s, redoing *all* the work instead of just the last bit becomes painful.

At the moment I work around this by deleting targeted fragments of `dist-newstyle/build`. This works, but:
- it's clumsy
- it's not discoverable
- it's probably not comprehensively cleaning the package (I have no idea if it gets all the files!)

I think it would be nice to provide a per package/component `cabal clean`, and hopefully it would just be a matter of listing all the relevant files and deleting them.

**Comment by Mikolaj** [comment]
I guess we'd need to delete the package or component files *and* the common files. The trouble is identifying those, in particular if they are not generated with `cabal build` but only with some other commands or with some specific options.

What does `cabal clean` do exactly? Only remove `dist-newstyle` or anything else?

**Comment by Mikolaj** [comment]
I see there is option

    -s, --save-config              Save configuration, only remove build artifacts

I wonder what's 'configuration' and 'artifacts' and if there is anything more. Code-diving would help.

**Comment by fendor** [comment]
> I see there is option
> 
> ```
> -s, --save-config              Save configuration, only remove build artifacts
> ```
> 
> I wonder what's 'configuration' and 'artifacts' and if there is anything more. Code-diving would help.

As far as I can tell, that flag is just being ignored, but the code path *should* delete `dist-newstyle/build` only which is still not fine grained enough for the proposed feature.

**Comment by fendor** [comment]
Looks rather intentional to me: https://github.com/haskell/cabal/blob/769e4ede066108cf722ac4828884788429c2d310/cabal-install/src/Distribution/Client/CmdClean.hs#L72 but the commit message itself gives no indication whether it really is intentional

**Comment by Mikolaj** [comment]
May it be that 'configuration' is only `cabal.project.local`, which is not removed by `cabal clean` and that's why `cabal clean` never removes configuration in the new v2- world? If so, we should mark the `-s` flag "deprecated; configuration is always saved now" once v1- commands are removed.

However, if  'configuration' is `cabal.project.local` and artifacts are `dist-newstyle/build`, is there anything else? E.g., any kind of caches? E.g., when downloading packages form repos?

**Comment by fendor** [comment]
I think currently `cabal clean` considers everything in `dist-newstyle/build` as build artefacts (which makes sense) and other files, e.g. `dist-newstyle/cache` is configuration. 

However, the help message from `Setup.hs` actually tells us roughly that:
```
-s --save-configure Do not remove the configuration file (dist/setup-config)
                     during cleaning. Saves need to reconfigure.
```
Not exactly the same flag but looks similar. Maybe `--save-config` is a remnant from `Setup.hs` that doesn't make much sense in cabal v2 world? I would argue at least in that sense.


Regarding the issue, I think there are two ways of tackling this:
* Calling Setup.hs clean for each component listed on the command line
* Basically call the functions from `Cabal` manually (assuming there are no Buildhooks) with the right parameters

**Comment by TeofilC** [comment]
As I mentioned on #7857, I'd be interested in trying to implement this.

On the other ticket:
  - @gbaz made the good point that when cleaning a component, we want to also clean anything that depends on it.
  - @bacchanalia mentioned this overlaps with their work on #7851. So, it seems to me like it would make sense to wait until that's merged to avoid conflicts

**Comment by bacchanalia** [comment]
@TeofilC #7851 has been merged, so you should be clear if you want to work on this.

**Comment by jneira** [comment]
`cabal clean` serves as a simple workaround for lot of problems hard to fix properly. Having a more fine grained cleaning would make the workaround more attractive
